streams/shared/src/main/scala/zio/stream/StreamEffect.scala

```diff
@@ -0,0 +1,289 @@
+    }
+
+  override def map[B](f0: A => B): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.map(f0).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal { () =>
+            f0(it())
+          }
+        }
+    }
+
+  override def mapAccum[S1, B](s1: S1)(f1: (S1, A) => (S1, B)): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.mapAccum(s1)(f1).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var state = s1
+
+            () => {
+              val (s2, b) = f1(state, it())
+              state = s2
+              b
+            }
+          }
+        }
+    }
+
+  override def mapConcat[B](f: A => Chunk[B]): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.mapConcat(f).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var chunk: Chunk[B] = Chunk.empty
+            var index           = 0
+
+            () => {
+              while (index == chunk.length) {
+                chunk = f(it())
+                index = 0
+              }
+              val b = chunk(index)
+              index += 1
+              b
+            }
+          }
+        }
+    }
+
+  override def run[R, E1 >: E, A0, A1 >: A, B](sink: ZSink[R, E1, A0, A1, B]): ZIO[R, E1, B] =
+    sink match {
+      case sink: SinkPure[E1, A0, A1, B] =>
+        foldEffect[ZSink.Step[sink.State, A0]](sink.initialPure)(ZSink.Step.cont) { (s, a) =>
+          sink.stepPure(ZSink.Step.state(s), a)
+        }.use[Any, E1, B] { step =>
+          ZIO.fromEither(sink.extractPure(ZSink.Step.state(step)))
+        }
+
+      case sink: ZSink[R, E1, A0, A1, B] => super.run(sink)
+    }
+
+  override def take(n: Int): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.take(n).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var counter = 0
+
+            () => {
+              if (counter >= n) StreamEffect.end
+              else {
+                counter += 1
+                it()
+              }
+            }
+          }
+        }
+    }
+
+  override def takeWhile(pred: A => Boolean): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.takeWhile(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal { () =>
+            {
+              val a = it()
+              if (pred(a)) a
+              else StreamEffect.end
+            }
+          }
+        }
+    }
+}
+
+private[stream] object StreamEffect extends Serializable {
+  import ZStream.InputStream
+
+  case class Failure[E](e: E) extends Throwable(e.toString, null, true, false)
```

Comment:
Do we need the `override def fillStacktrace() = this` trick on these two?

Reply:
The original code had that, but then I switched to the code snippet that @jdegoes provided.

Relevant stackoverflow [link](https://stackoverflow.com/questions/11434431/exception-without-stack-trace-in-java).

It looks like we are good.

Reply:
Done.